designdocs/permissions.md

```diff
@@ -0,0 +1,87 @@
+* or are the permissions granted at the time of use.
+
+### stating concerns
+
+Concerns about granting at page load time are
+* users may be requested to approve multiple permissions one after another, creating click fatigue.
+* if all apps request permissions at start, the user has no information to judge whether or not they should grant this, because they haven’t entered the experience yet.
+
+Concerns about granting requests once already inside of immersive mode are:
+* access requests when inside immersive mode should be done with some sort of a secure method, such as a dedicated hardware button, so that applications cannot spoof such permissions.
+* some devices or scenarios may not be able to provide such a secure method, thus doing the request before entering immersive mode is preferable.
+
+Concerns about both systems include:
+* the user agents know more about the user, the current scenario, and can make better decisions *at the time of use* than a spec can. Neither method give the user agent freedom to adapt.
+* there may be new attacks devised or new flaws discovered after we have shipped, but user agents will not be able to address these because they are constrained by the spec.
+
+There are most likely more concerns than just what I have listed above.
+
+### Solution
+To address these I propose the following: *do both*.  Let the developers have a single code path where they /request permissions at application start/ *and* when /the permission is actually needed/. The first request is considered the upper bounds of possible permissions. The user agent can decide whether to actually show user request dialogs at app time or when the permission is actually used, or some other scheme that the user agent deems to be better (such as warding off new attacks in the future).
+
+### Details
+From the developer’s perspective it would work like this.
+
+The developer specifies the total permissions they are likely to need during the entire run of the application when requesting the XR Session, using an array with pre-defined permissions constants.  Something like this:
+
+``` javascript
+const perms = [
+   MICROPHONE,
+	 IMMERSIVE_MODE,
+   HEADSET_POSE,
+   //CAMERA_DATA,  developer decides not to use camera
+]
+navigator.xr.requestSession(perms).then(session => {
+	//start up the rest of the app
+})
+```
+
+At this point the user agent can either prompt the user for these permissions, or merely use them as an upper bound of permissions that could potentially be requested later.
+
+
+Later, when the application needs to actually use a permission, say the microphone, the developer will request this access through a promise based API. Ex:
+
+```javascript
+navigator.mediaDevices.getUserMedia(opts).then(stream => {
+	console.log("i have the microphone stream")
+})
+```
+
+At this point the user agent can either prompt the user for the microphone permission, or immediately grant it if the user already approved it previously.  
```

Comment:
One caveat to this is that the timing of the application matters, and that cannot be known by the user agent. For example: Let's assume I have two applications that both use some plane geometry and camera access. _AppA_ is an app where you place virtual furniture, and can get away with only having plane access until the user says they want to save a picture. _AppB_ is tour guide app and both plane access and camera access are absolutely essential to it's function.

If we assume that the UA is capable of doing just-in-time prompts, _AppA_ works out pretty well, in that the app starts up and asks "Planes, please?" Then waits until the user clicks the "Save image" button at which point the user is asked "Camera access: Cool, or nah?" User understands the context, clicks "Yes", everyone is happy.

However, with _AppB_ the user first immediately sees a prompt saying "Planes!" then, upon clicking yes or no gets an immediate followup prompt says "Camera!" and now we're just pissing them off by making them click through multiple layers of prompts because the UA wanted to do the "right thing" of deferring permission till the time it was needed.

As such, we probably want the developer to be able to specify a list of permissions that MUST be requested prior to session creation, as well as permissions they MAY use later (which could then be requested up front based on UA capabilities if needed.)

Reply:
If we do take the approach of integrating with https://www.w3.org/TR/permissions/#enumdef-permissionname then we can also take advantage of the https://www.w3.org/TR/permissions/#dom-permissions-query API?  Which to be honest, is interesting to me that this is queryable since for some reason I thought there were concerns about websites being able to differentiate between a user rejecting a permission and some other random reason a feature might not be available?  🤔

Reply:
It's funny.. This exact thing was something I used to be pretty worried about... But this is how the rest of the web works today, yeah?  So I'm kinda less worried?  A couple things to consider:
* Even if we manage to do something like you've described for XR related permissions, the same problem can occur for non-XR permission prompts while in an immersive session.  Who's to say a developer won't ask for microphone and the immediately ask for geolocation? 
* Alternatively, given that promises aren't resolved immediately, perhaps user agents have the option to be clever and also delay the display of a permission prompt until after control has returned to the user agent?  In which case a developer invoking two permissions requests in the same function flow would cause that consolidation to occur, yes?

Reply:
Is this the same thing @johnpallett was asking about above. A way to specify if a particular permission is must vs may (required vs optional)?

Reply:
I'd actually asked whether there should be three options for feature requests: "required", "optional" and "later, not now" (the third one I don't have the right name for). @joshmarinacci this was based upon what I thought you described during the f2f.

The idea behind these three options as I understood them were:
_Required_: I need this. Otherwise, fail to create the session.
_Optional_: I don't necessarily need this, but please get consent now so I know what I have.
_Later, Not Now_: I don't necessarily need this, and I don't need to know now. You can ask later, or now, either works for me.

That way the user agent has a sense of when to ask for user consent: Required at session creation, Optional at session creation (if the platform supports the feature), 'Later not now' either at session creation, or when the feature is actually used, the choice being at the discretion of the user agent.